keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,739 @@
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+ Add `NamespaceSelector` field to `PodAffinityTerm`:
+
+```go
+type PodAffinityTerm struct {
+    // A label query over the set of namespaces that the term applies to.
+	// The term is applied to the union of the namespaces selected by this field
+	// and the ones listed in the namespaces field.
+	// nil selector and nil or empty namespaces list means "this pod's namespace"
+	// The only valid operator is In.
+    NamespaceSelector *metav1.LabelSelector
+}
+```
+
+As indicated in the comment, the scheduler will consider the union of the namespaces
+specified in the existing `Namespaces` field and the ones selected by the new 
+`NamespaceSelector` field. `NamespaceSelector`  is ignored when set to nil.
+
+We will do two precomputations at PreFilter/PreScore:
+
+- The names of the namespaces selected by the `NamespaceSelector` is computed. This set
+  will be used by Fliter/Score to match against existing pods namespaces.
+- A snapshot of the namespace labels of the incoming pod. This will be used when 
+  to match against the anti-affinity constraints of existing pods. 
+
+The precomputations are necessary for:
+
+- Performance.
+- Ensures a consistent behavior if namespace labels are added/removed during 
+  the scheduling cycle of a pod.
+
+Finally, the feature will be guarded by a new feature flag. If the feature is 
+disabled, the field `NamespaceSelector` is preserved if it was already set in
+the persisted Pod ojbect, otherwise it is silently dropped; moreover kube-scheduler 
+will ignore the field and continue to behave as before.
+
+
+With regards to adding the `CrossNamespaceAffinity` quota scope, the one design aspect
+worth noting is that it will be rolled out in multiple releases similar to the `NamespaceSelector`:
+when the feature is disabled, the new value will be tolerated in updates of objects
+already containing the new value, but will not be allowed to be added on create or update.
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+- Unit and integration tests covering:
+  - core changes
+  - correctness for namespace addition/removal, specifically label updates 
+    should be taken into account, but not in the middle of a scheduling cycle
+  - feature gate enabled/disabled
+- Benchmark Tests: 
+  - evaluate performance for the case where the selector matches a large number of namespaces
```

Comment:
In particular, I would be interested in seeing a test, where most of pods would be covered (e.g. all namespaces by kube-system or sth like that).

Reply:
added large number of pods explicitly. The plan is to have every pod with affinity to every other pod (a test we already have, but in a single namespace).